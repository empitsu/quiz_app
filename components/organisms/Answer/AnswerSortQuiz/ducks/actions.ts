import { SortableOptionsSet } from "./model";
import {
  POP_FROM_SELECTED_OPTIONS,
  PUSH_T0_SELECTED_OPTIONS,
  RESET,
} from "./types";

export const popFromSelectedOptions = (
  removingOption: NonNullable<SortableOptionsSet["selectedOptions"][number]>
) =>
  ({
    type: POP_FROM_SELECTED_OPTIONS,
    payload: removingOption,
  } as const);

export const pushToSelectedOptions = (
  addingOption: NonNullable<SortableOptionsSet["selectedOptions"][number]>
) =>
  ({
    type: PUSH_T0_SELECTED_OPTIONS,
    payload: addingOption,
  } as const);

export const reset = (originalOptions: SortableOptionsSet["restOptions"]) =>
  ({
    type: RESET,
    payload: originalOptions,
  } as const);

export type Actions =
  | ReturnType<typeof popFromSelectedOptions>
  | ReturnType<typeof pushToSelectedOptions>
  | ReturnType<typeof reset>;
