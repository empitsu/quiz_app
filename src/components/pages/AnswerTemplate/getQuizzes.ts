import { getApp } from "../../../utils/firebaseHelpers";

// TODO:共通化
type Option = {
  optionId: number;
  text: string;
};

type Quiz =
  | {
      type: "sort";
      title: string;
      options: Option[];
    }
  | {
      type: "selection";
      title: string;
      correctOptionId: number;
      options: Option[];
    };

export type QuizArray = {
  id: string;
  data: Quiz;
}[];

export async function getQuizzes(): Promise<QuizArray> {
  const auth = getApp().auth();
  if (!auth || auth.currentUser === null) {
    throw new Error("ログインされていません");
  }

  const token = await auth.currentUser.getIdToken(true);
  try {
    const res = await fetch(`/api/quizzes`, {
      method: "GET",
      headers: {
        authorization: token ? `Bearer ${token}` : "",
      },
    });
    console.log("res", res);

    const responseJson = await res.json();

    console.log("responseJson", responseJson);
    if (res.ok) {
      return responseJson;
    } else {
      const errorType: string = ((responseJson as unknown) as { type: string })
        .type;

      throw new Error(errorType);
    }
  } catch (error) {
    throw new Error(error);
  }
}
