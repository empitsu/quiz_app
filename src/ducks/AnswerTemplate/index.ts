export { answerProps } from "./model";
export { incrementCurrentQuiz } from "./actions";
export { reducer } from "./reducer";
