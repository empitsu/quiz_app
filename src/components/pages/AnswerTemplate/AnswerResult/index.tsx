import { useContext } from "react";
import { AnswerPropStore } from "../../../../contexts/AnswerProps";

import { incrementCurrentQuiz } from "../../../../ducks/AnswerTemplate";
import { FiberManualRecordOutlined, Clear } from "@material-ui/icons";
import { Button } from "../../../uikit/Button";

import styled from "styled-components";

type ResultProps = {
  isCorrect: boolean;
  officialAnswer: string;
};

const StyledFiberManualRecordOutlinedIcon = styled(FiberManualRecordOutlined)`
  && {
    font-size: 150px;
    color: ${({ theme }) => theme.palettes.success.main};
  }
`;

const StyledClear = styled(Clear)`
  && {
    font-size: 150px;
    color: ${({ theme }) => theme.palettes.error.main};
  }
`;
const StyledWrapDiv = styled.div`
  text-align: center;
  width: 100%;
`;

export function AnswerResult({ isCorrect, officialAnswer }: ResultProps) {
  const { dispatch } = useContext(AnswerPropStore);

  if (isCorrect) {
    return (
      <StyledWrapDiv>
        <div>
          <StyledFiberManualRecordOutlinedIcon></StyledFiberManualRecordOutlinedIcon>
        </div>

        <p>正解！</p>
        <p>正解：{officialAnswer}</p>
        <Button
          isFullWidth
          color="info"
          onClick={() => {
            dispatch(incrementCurrentQuiz(true));
          }}
        >
          次へ
        </Button>
      </StyledWrapDiv>
    );
  }
  return (
    <StyledWrapDiv>
      <div>
        <StyledClear></StyledClear>
      </div>
      <p>不正解！</p>
      <p>正解：{officialAnswer}</p>

      <Button
        isFullWidth
        color="info"
        onClick={() => {
          dispatch(incrementCurrentQuiz(false));
        }}
      >
        次へ
      </Button>
    </StyledWrapDiv>
  );
}
