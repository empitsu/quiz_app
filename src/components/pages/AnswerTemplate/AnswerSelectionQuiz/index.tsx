import { useCallback, useState } from "react";
import styled from "styled-components";
import { Button } from "../../../uikit/Button";
import { Heading } from "../../../uikit/Heading";
import { AnswerResult } from "../AnswerResult";
// todo:共通化
type Option = {
  optionId: number;
  text: string;
};

type AnswerQuizProps = {
  title: string;
  correctOptionId: number;
  options: Option[];
};

type ButtonToAnswerProps = {
  optionId: number;
  text: string;
  correctOptionId: number;
  onAnswerCorrectly: () => void;
  onMistake: () => void;
};

function ButtonToAnswer({
  optionId,
  text,
  correctOptionId,
  onAnswerCorrectly,
  onMistake,
}: ButtonToAnswerProps) {
  const onClickAnswer = useCallback(() => {
    if (optionId === correctOptionId) {
      onAnswerCorrectly();
    } else {
      onMistake();
    }
  }, [correctOptionId, onAnswerCorrectly, onMistake, optionId]);

  return (
    <Button isFullWidth color="secondary" onClick={onClickAnswer}>
      {text}
    </Button>
  );
}

const StyledQuizTitleP = styled.p`
  text-align: center;
  margin-bottom: 30px;
`;

const StyledList = styled.li`
  list-style: none;
  margin-bottom: 10px;
`;

const StyledUl = styled.ul`
  margin-bottom: 50px;
  padding: 0;
`;

const StyledSelectionQuizWrapDiv = styled.div`
  position: relative;
`;

export function AnswerSelectionQuiz({
  title,
  options,
  correctOptionId,
}: AnswerQuizProps) {
  const officialAnswer = options.find(
    (option) => option.optionId === correctOptionId
  )?.text;
  const [isCorrect, setIsCorrect] = useState<boolean | null>(null);
  const onAnswerCorrectly = useCallback(() => {
    setIsCorrect(true);
  }, []);
  const onMistake = useCallback(() => {
    setIsCorrect(false);
  }, []);
  return (
    <article>
      <Heading styleLevel="h3">選択してください</Heading>

      <StyledQuizTitleP>{title}</StyledQuizTitleP>
      <StyledSelectionQuizWrapDiv>
        {isCorrect === null ? (
          <StyledUl>
            {options.map((option) => {
              return (
                <StyledList key={option.optionId}>
                  <ButtonToAnswer
                    optionId={option.optionId}
                    text={option.text}
                    correctOptionId={correctOptionId}
                    onAnswerCorrectly={onAnswerCorrectly}
                    onMistake={onMistake}
                  ></ButtonToAnswer>
                </StyledList>
              );
            })}
          </StyledUl>
        ) : (
          <AnswerResult
            isCorrect={isCorrect}
            officialAnswer={officialAnswer ?? ""}
          ></AnswerResult>
        )}
      </StyledSelectionQuizWrapDiv>
    </article>
  );
}
