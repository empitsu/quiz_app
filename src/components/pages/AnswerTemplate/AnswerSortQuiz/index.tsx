import { Dispatch, useCallback, useReducer, useState } from "react";
import { reducer } from "../../../../ducks/AnswerSortQuiz";
import {
  Actions,
  popFromSelectedOptions,
  pushToSelectedOptions,
} from "../../../../ducks/AnswerSortQuiz/actions";
import { SortableOptionsSet } from "../../../../ducks/AnswerSortQuiz/model";
import { Button } from "../../../uikit/Button";
import styled from "styled-components";
import { shuffleArray } from "../../../../utils/shuffleArray";
import { AnswerResult } from "../AnswerResult";
import { Heading } from "../../../uikit/Heading";

type AnswerQuizProps = {
  title: string;
  options: {
    optionId: number;
    text: string;
  }[];
};

type EachOptionButtonProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

const StyledEmptyOptionButton = styled(Button)`
  color: transparent;
`;

function EachOptionButton({ option, dispatch }: EachOptionButtonProps) {
  const onClickAnswer = useCallback(() => {
    dispatch(pushToSelectedOptions(option));
  }, [dispatch, option]);

  if (option.selected) {
    return (
      <StyledEmptyOptionButton disabled>{option.text}</StyledEmptyOptionButton>
    );
  }

  return (
    <Button color="secondary" onClick={onClickAnswer}>
      {option.text}
    </Button>
  );
}

type SelectedOptionProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

function SelectedOption({ option, dispatch }: SelectedOptionProps) {
  const onClickSelectedOption = useCallback(() => {
    dispatch(popFromSelectedOptions(option));
  }, [dispatch, option]);

  return (
    <Button color="secondary" onClick={onClickSelectedOption}>
      {option.text}
    </Button>
  );
}

const StyledQuizTitleP = styled.p`
  text-align: center;
  margin-bottom: 30px;
`;

const StyledSelectedOptionUl = styled.ul`
  min-height: 106px;
  display: flex;
  flex-wrap: wrap;
  padding: 30px 0;
  border-top: 1px solid ${({ theme }) => theme.palettes.grey["600"]};
  border-bottom: 1px solid ${({ theme }) => theme.palettes.grey["600"]};
  margin: 30px 0;
`;

const StyledRestOptionUl = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 30px 0;
  margin-bottom: 50px;
`;
const StyledList = styled.li`
  list-style: none;
  margin-right: 5px;
`;

const StyledSelectionQuizWrapDiv = styled.div`
  position: relative;
`;

export function AnswerSortQuiz({ title, options }: AnswerQuizProps) {
  const officialAnswer = [...options]
    .sort()
    .map((option) => option.text)
    .join(" ");
  const defaultRestOptions = shuffleArray(options).map((option, index) => {
    return {
      ...option,
      originalIndex: index,
      selected: false,
    };
  });
  const [state, dispatch] = useReducer(reducer, {
    selectedOptions: [],
    restOptions: [...defaultRestOptions],
  });

  const [isCorrect, setIsCorrect] = useState<boolean | null>(null);

  // 正解かどうかチェックする
  const onClickAnswerBtn = useCallback(() => {
    const isCorrect =
      state.selectedOptions.length === 0
        ? false
        : state.selectedOptions.every((option, index) => {
            return option.optionId === index;
          });

    setIsCorrect(isCorrect);
  }, [state.selectedOptions]);
  return (
    <article>
      <Heading styleLevel="h3">ボタンを順番に選択してください</Heading>
      <StyledQuizTitleP>{title}</StyledQuizTitleP>
      <StyledSelectionQuizWrapDiv>
        {isCorrect === null ? (
          <div>
            <StyledSelectedOptionUl>
              {state.selectedOptions.map((option) => {
                return (
                  <StyledList key={option.optionId}>
                    <SelectedOption
                      option={option}
                      dispatch={dispatch}
                    ></SelectedOption>
                  </StyledList>
                );
              })}
            </StyledSelectedOptionUl>
            <StyledRestOptionUl>
              {state.restOptions.map((option) => {
                return (
                  <StyledList key={option.optionId}>
                    <EachOptionButton
                      option={option}
                      dispatch={dispatch}
                    ></EachOptionButton>
                  </StyledList>
                );
              })}
            </StyledRestOptionUl>
            <Button
              disabled={isCorrect !== null}
              isFullWidth
              color="secondary"
              onClick={onClickAnswerBtn}
            >
              これで回答する
            </Button>
          </div>
        ) : (
          <AnswerResult
            isCorrect={isCorrect}
            officialAnswer={officialAnswer}
          ></AnswerResult>
        )}
      </StyledSelectionQuizWrapDiv>
    </article>
  );
}
