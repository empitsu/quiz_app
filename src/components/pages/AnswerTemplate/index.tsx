import Link from "next/link";
import { useCallback, useContext, useEffect, useRef, useState } from "react";
import { getQuizzes, QuizArray } from "./getQuizzes";
import { useRouter } from "next/router";
import { AnswerPropStore } from "../../../contexts/AnswerProps";
import { Column1 } from "../../uikit/Column1";
import { Heading } from "../../uikit/Heading";
import { FormErrorText } from "../../uikit/FormErrorText";
import styled from "styled-components";
import { Button } from "../../uikit/Button";
import { FormItemWrap } from "../../uikit/FormItemWrap";
import { LinkText } from "../../uikit/LinkText";
import { AnswerSortQuiz } from "./AnswerSortQuiz";
import { AnswerSelectionQuiz } from "./AnswerSelectionQuiz";

const StyledNoQuizP = styled.p`
  text-align: center;
`;

const StyledResultP = styled.p`
  text-align: center;
  margin: 60px 0;
  font-size: 1.5rem;
`;

// TODO: quizzesLengthはcontextから取得するようにする。
function OverallGrade({ quizzesLength }: { quizzesLength: number }) {
  const { state } = useContext(AnswerPropStore);

  const router = useRouter();
  const onClickRetryBtn = useCallback(() => {
    router.reload();
  }, [router]);
  return (
    <Column1 size="small">
      <StyledResultP>
        {quizzesLength}問中{state.correctAnswersLength}問正解しました。
      </StyledResultP>
      <FormItemWrap>
        <Button isFullWidth color="primary" onClick={onClickRetryBtn}>
          もう一度挑戦
        </Button>
      </FormItemWrap>
      <FormItemWrap>
        <Link href="/mypage/">
          <Button isFullWidth color="secondary">
            TOPに戻る
          </Button>
        </Link>
      </FormItemWrap>
    </Column1>
  );
}

const StyledCancelP = styled.p`
  text-align: center;
  margin: 20px 0;
`;

// TODO: quiz dataやidもcontextに保存する。テストが書きやすくなるため。
function AnswerSelectionOrSortQuiz({ quiz }: { quiz: QuizArray[number] }) {
  const router = useRouter();
  const onClickCancelBtn = useCallback(() => {
    router.reload();
  }, [router]);
  if (quiz.data.type === "selection") {
    return (
      <>
        <AnswerSelectionQuiz
          key={quiz.id}
          title={quiz.data.title}
          options={quiz.data.options}
          correctOptionId={quiz.data.correctOptionId}
        ></AnswerSelectionQuiz>
        <StyledCancelP>
          <LinkText onClick={onClickCancelBtn}>キャンセルしてやり直す</LinkText>
        </StyledCancelP>
      </>
    );
  }

  return (
    <>
      <AnswerSortQuiz
        key={quiz.id}
        title={quiz.data.title}
        options={quiz.data.options}
      ></AnswerSortQuiz>
      <StyledCancelP>
        <LinkText onClick={onClickCancelBtn}>キャンセルしてやり直す</LinkText>
      </StyledCancelP>
    </>
  );
}

export default function AnswerTemplate() {
  const { state } = useContext(AnswerPropStore);
  const [error, setError] = useState<Error | null>(null);
  const [docs, setDocs] = useState<null | QuizArray>(null);
  const isMountedRef = useRef<null | boolean>(null);

  const isEmptyQuiz = docs !== null && docs.length === 0;
  const isQuizAvailable = docs !== null && state.currentQuizIndex < docs.length;
  const isShowResult =
    docs !== null && docs.length !== 0 && state.currentQuizIndex >= docs.length;

  useEffect(() => {
    isMountedRef.current = true;
    (async () => {
      try {
        // TODO: useEffectごとカスタムフックス化。hooksの中でcontextに保存するとテストも書きやすくなる。
        const docs = await getQuizzes();
        setDocs(docs);
      } catch (error) {
        setError(error);
      }
    })();
    return () => {
      isMountedRef.current = false;
    };
  }, []);
  return (
    <Column1 size="small">
      <Heading styleLevel="h2">クイズに挑戦</Heading>
      {error && (
        <FormErrorText role="alert" aria-label="通信時エラー">
          {error.message}
        </FormErrorText>
      )}
      {isEmptyQuiz && (
        <StyledNoQuizP>
          まだ問題がありません。登録画面からクイズを登録してください。
        </StyledNoQuizP>
      )}
      {isQuizAvailable && (
        <AnswerSelectionOrSortQuiz
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          quiz={docs![state.currentQuizIndex]}
        />
      )}
      {isShowResult && (
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        <OverallGrade quizzesLength={docs!.length}></OverallGrade>
      )}
    </Column1>
  );
}
