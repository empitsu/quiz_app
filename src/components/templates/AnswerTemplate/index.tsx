import type firebase from "firebase";
import Link from "next/link";

import { useCallback, useContext, useEffect, useRef, useState } from "react";

import { getQuizzes } from "../../../utils/getQuizzes";
import { AnswerSelectionQuiz } from "../../../projects/Answer/AnswerSelectionQuiz";
import { AnswerSortQuiz } from "../../../projects/Answer/AnswerSortQuiz";
import { useRouter } from "next/router";
import { AnswerPropStore } from "../../../contexts/AnswerProps";
import { Column1 } from "../../atoms/Column1";
import { Heading } from "../../atoms/Heading";
import { FormErrorText } from "../../atoms/FormErrorText";
import styled from "styled-components";
import { Button } from "../../atoms/Button";
import { FormItemWrap } from "../../atoms/FormItemWrap";
import { LinkText } from "../../atoms/LinkText";

const StyledNoQuizP = styled.p`
  text-align: center;
`;

const StyledResultP = styled.p`
  text-align: center;
  margin: 60px 0;
  font-size: 1.5rem;
`;

// TODO: quizzesLengthはcontextから取得するようにする。
function OverallGrade({ quizzesLength }: { quizzesLength: number }) {
  const { state } = useContext(AnswerPropStore);

  const router = useRouter();
  const onClickRetryBtn = useCallback(() => {
    router.reload();
  }, [router]);
  return (
    <Column1 size="small">
      <StyledResultP>
        {quizzesLength}問中{state.correctAnswersLength}問正解しました。
      </StyledResultP>
      <FormItemWrap>
        <Button isFullWidth color="secondary" onClick={onClickRetryBtn}>
          もう一度挑戦
        </Button>
      </FormItemWrap>
      <FormItemWrap>
        <Link href="/mypage/">
          <Button isFullWidth color="info">
            TOPに戻る
          </Button>
        </Link>
      </FormItemWrap>
    </Column1>
  );
}

const StyledCancelP = styled.p`
  text-align: center;
  margin: 20px 0;
`;

// TODO: quiz dataやidもcontextに保存する。テストが書きやすくなるため。
function AnswerSelectionOrSortQuiz({
  quizId,
  quiz,
}: {
  quizId: string;
  quiz: QuizData;
}) {
  const router = useRouter();
  const onClickCancelBtn = useCallback(() => {
    router.reload();
  }, [router]);
  if (quiz.type === "selection") {
    return (
      <>
        <AnswerSelectionQuiz
          key={quizId}
          title={quiz.title}
          options={quiz.options}
          correctOptionId={quiz.correctOptionId}
        ></AnswerSelectionQuiz>
        <StyledCancelP>
          <LinkText onClick={onClickCancelBtn}>キャンセルしてやり直す</LinkText>
        </StyledCancelP>
      </>
    );
  }

  return (
    <>
      <AnswerSortQuiz
        key={quizId}
        title={quiz.title}
        options={quiz.options}
      ></AnswerSortQuiz>
      <StyledCancelP>
        <LinkText onClick={onClickCancelBtn}>キャンセルしてやり直す</LinkText>
      </StyledCancelP>
    </>
  );
}

// todo:共通化
export type QuizData =
  | {
      type: "sort";
      title: string;
      options: {
        optionId: number;
        text: string;
      }[];
    }
  | {
      type: "selection";
      title: string;
      correctOptionId: number;
      options: {
        optionId: number;
        text: string;
      }[];
    };

//
type Docs = firebase.firestore.QueryDocumentSnapshot<firebase.firestore.DocumentData>[];
export default function AnswerTemplate() {
  const { state } = useContext(AnswerPropStore);
  const [error, setError] = useState<Error | null>(null);
  const [docs, setDocs] = useState<null | Docs>(null);
  const isMountedRef = useRef<null | boolean>(null);

  useEffect(() => {
    isMountedRef.current = true;
    (async () => {
      try {
        // TODO: useEffectごとカスタムフックス化。hooksの中でcontextに保存するとテストも書きやすくなる。
        const docs = await getQuizzes();
        setDocs(docs);
      } catch (error) {
        setError(error);
      }
    })();
    return () => {
      isMountedRef.current = false;
    };
  }, []);
  return (
    <Column1 size="small">
      <Heading styleLevel="h2">クイズに挑戦</Heading>
      {error && (
        <FormErrorText role="alert" aria-label="通信時エラー">
          {error.message}
        </FormErrorText>
      )}
      {docs !== null && docs.length === 0 && (
        <StyledNoQuizP>
          まだ問題がありません。登録画面からクイズを登録してください。
        </StyledNoQuizP>
      )}
      {docs !== null && state.currentQuizIndex < docs.length && (
        <AnswerSelectionOrSortQuiz
          quizId={docs[state.currentQuizIndex].id}
          quiz={docs[state.currentQuizIndex].data() as QuizData}
        />
      )}
      {docs !== null && state.currentQuizIndex >= docs.length && (
        <OverallGrade quizzesLength={docs.length}></OverallGrade>
      )}
    </Column1>
  );
}
