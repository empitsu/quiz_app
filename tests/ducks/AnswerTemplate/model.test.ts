import { answerProps } from "../../../src/ducks/AnswerTemplate";

describe("answerProps", () => {
  it("should create a empty model", () => {
    expect(answerProps).toMatchSnapshot();
  });
});
