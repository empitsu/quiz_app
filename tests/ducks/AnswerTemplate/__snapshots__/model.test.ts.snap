// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`answerProps should create a empty model 1`] = `
Object {
  "correctAnswersLength": 0,
  "currentQuizIndex": 0,
}
`;
